core/shared/src/main/scala/zio/Semaphore.scala

```diff
@@ -18,153 +18,173 @@ package zio
-
-        def withPermitsScoped(n: Long)(implicit trace: Trace): ZIO[Scope, Nothing, Unit] =
-          ZIO.acquireRelease(reserve(n))(_.release).flatMap(_.acquire)
-
-        case class Reservation(acquire: UIO[Unit], release: UIO[Any])
-
-        def reserve(n: Long)(implicit trace: Trace): UIO[Reservation] =
-          if (n < 0)
-            ZIO.die(new IllegalArgumentException(s"Unexpected negative `$n` permits requested."))
-          else if (n == 0L)
-            ZIO.succeedNow(Reservation(ZIO.unit, ZIO.unit))
-          else
-            Promise.make[Nothing, Unit].flatMap { promise =>
-              ref.modify {
-                case Right(permits) if permits >= n =>
-                  Reservation(ZIO.unit, releaseN(n)) -> Right(permits - n)
-                case Right(permits) =>
-                  Reservation(promise.await, restore(promise, n)) -> Left(ScalaQueue(promise -> (n - permits)))
-                case Left(queue) =>
-                  Reservation(promise.await, restore(promise, n)) -> Left(queue.enqueue(promise -> n))
-              }
-            }
+    def make(permits: Long, fairness: Boolean)(implicit unsafe: Unsafe): Semaphore = {
+      if (fairness) new FairSemaphore(permits)
+      else new UnfairSemaphore(permits)
+    }
+  }
 
-        def restore(promise: Promise[Nothing, Unit], n: Long)(implicit trace: Trace): UIO[Any] =
-          ref.modify {
-            case Left(queue) =>
-              queue
-                .find(_._1 == promise)
-                .fold(releaseN(n) -> Left(queue)) { case (_, permits) =>
-                  releaseN(n - permits) -> Left(queue.filter(_._1 != promise))
+  private final class FairSemaphore(initialPermits: Long) extends Semaphore {
+    private val state = new AtomicReference[Either[ScalaQueue[(Promise[Nothing, Unit], Long)], Long]](Right(initialPermits))
+
+    def available(implicit trace: Trace): UIO[Long] =
+      ZIO.succeed(state.get() match {
+        case Left(_)        => 0L
+        case Right(permits) => permits
+      })
+
+    def awaiting(implicit trace: Trace): UIO[Long] =
+      ZIO.succeed(state.get() match {
+        case Left(queue) => queue.size.toLong
+        case Right(_)    => 0L
+      })
+
+    def withPermit[R, E, A](zio: ZIO[R, E, A])(implicit trace: Trace): ZIO[R, E, A] =
+      withPermits(1L)(zio)
+
+    def withPermitScoped(implicit trace: Trace): ZIO[Scope, Nothing, Unit] =
+      withPermitsScoped(1L)
+
+    def withPermits[R, E, A](n: Long)(zio: ZIO[R, E, A])(implicit trace: Trace): ZIO[R, E, A] =
+      ZIO.acquireReleaseWith(reserve(n))(_.release)(_.acquire *> zio)
+
+    def withPermitsScoped(n: Long)(implicit trace: Trace): ZIO[Scope, Nothing, Unit] =
+      ZIO.acquireRelease(reserve(n))(_.release).flatMap(_.acquire)
+
+    def releaseN(n: Long)(implicit trace: Trace): UIO[Unit] =
+      ZIO.succeed(releaseInternal(n))
+
+    private def reserve(n: Long)(implicit trace: Trace): UIO[Reservation] = {
```

Comment:
Where is `Reservation` defined now? 

Should it be a `protected` class in the `Semphore` trait, or package private in the`Semaphore` object?